Horizontal read scaling (#4028)
timescale/docs: use-timescale/ha-replicas/read-scaling.md

You use $READ_REPLICA sets for horizontal **read** scaling. To limit data loss for your $SERVICE_LONGs, use [$HA_REPLICAs][ha].

## Prerequisites

To follow this procedure:

- Create a target $SERVICE_LONG.
- Create a [read-only user][read-only-role] on the primary data instance. 

  A user with read-only permissions cannot make changes in the primary database. This user is propagated to the $READ_REPLICA set when you create it.

## Create a $READ_REPLICA set

To create a secure $READ_REPLICA set for your read-intensive apps: 

<Procedure>

1. **In [$CONSOLE][timescale-console-services], select your target $SERVICE_SHORT**

1. **Click `Operations` > `Read scaling` > `Add a read replica set`**

1. **Configure your replica set** 

    Configure the number of nodes, compute size, connection pooling, and the name for your replica, then click `Create read replica set`.

   ![Create a read replica set in Timescale Console](https://assets.timescale.com/docs/images/create-read-replica-set-timescale-console.png)

1. **Save the connection information**

   The username and password of a read replica set are the same as the primary $SERVICE_SHORT. They cannot be changed independently. 

>    The connection information for each $READ_REPLICA set is unique. If you add or remove nodes from an existing set, the connection information of that set changes. To find the connection information for an existing $READ_REPLICA set: 

Reviewer: ```suggestion
   The connection information for each $READ_REPLICA set is unique. You can add or remove nodes from an existing set and the connection information of that set will remain the same. To find the connection information for an existing $READ_REPLICA set: 
```

Author: @ramonguiu just checking: this was originally in the PRD.  I have tested this on dev and it seems the connection string did change when changing the number of nodes?

     1. Select the primary $SERVICE_SHORT in $CONSOLE.
    
     1. Click `Operations` > `Read scaling`.

     1. Click the 🔗 icon next to the replica set in the list. 

    

</Procedure>

## Edit a $READ_REPLICA set

You can edit an existing $READ_REPLICA set to better handle your reads. This includes changing the number of nodes, compute size, storage, and IOPS, as well as configuring $VPC and other features. 

<Procedure>

To change the compute and storage configuration of your $READ_REPLICA set: 

1. **In [$CONSOLE][timescale-console-services], expand and click the $READ_REPLICA set under your primary $SERVICE_SHORT**

   ![Read replicas in Timescale Console](https://assets.timescale.com/docs/images/read-replica-sets-timescale-console.png)

1. **Click `Operations` > `Compute and storage`**

   ![Read replica compute and storage in Timescale Console](https://assets.timescale.com/docs/images/read-replica-set-configs.png)

1. **Change the replica configuration and click `Apply`**

>    If you add or remove nodes from an existing $READ_REPLICA set, reconnect to the set using the updated connection details. Find the connection details 

Reviewer: ```suggestion
```

---
title: Read scaling
excerpt: For read-intensive apps, Timescale Cloud enables you to create read-only replica sets that take over read queries and offload your primary node. Create read-only replica sets with automated load balancing in Timescale Console
product: cloud
price_plans: [scale, enterprise]
keywords: [replicas, scaling]
tags: [replicas, scaling, ha]
---

# Read scaling

> You use $READ_REPLICA sets in $CLOUD_LONG for horizontal read scaling, to power your read-intensive apps and business intelligence tooling. Using the replica sets to serve reads for your app unloads the primary data instance and enables your $SERVICE_SHORT to improve ingest performance. 

Reviewer: ```suggestion
You use $READ_REPLICA sets in $CLOUD_LONG for horizontal read scaling, to power your read-intensive apps and business intelligence tooling. Additionally, using the replica sets to serve reads for your app unloads the primary data instance and enables your $SERVICE_SHORT to improve ingest performance. 
```

This is particularly useful when read traffic is very spiky and risks impacting ingest performance, or where reads have 
a lower priority to writes. 

This page shows you how to create and manage $READ_REPLICA sets in $CONSOLE.

## What is read replication?

> A $READ_REPLICA is a read-only copy of your primary data instance. Queries on $READ_REPLICA have minimal impact on the performance of the primary instance. This enables you to interact with up-to-date production data for analysis, or to scale out reads beyond the limits of your primary instance. $READ_REPLICA_CAPs can be short-lived and deleted when a session of data analysis is complete, or long-running to power a business intelligence tool. 

Reviewer: ```suggestion
A $READ_REPLICA is a read-only copy of your primary database instance. Queries on read replicas have minimal impact on the performance of the primary instance. This enables you to interact with up-to-date production data for analysis, or to scale out reads beyond the limits of your primary instance. $READ_REPLICA_CAPs can be short-lived and deleted when a session of data analysis is complete, or long-running to power an application or a business intelligence tool. 
```

> A $READ_REPLICA set in $CLOUD_LONG is a group of one or more $READ_REPLICA nodes that share the same endpoint. You query each set as a single replica. $CLOUD_LONG balances the load between the nodes in the set for you.

Reviewer: ```suggestion
A $READ_REPLICA set in $CLOUD_LONG is a group of one or more $READ_REPLICA nodes that are accessed through the same endpoint. You query each set as a single replica. $CLOUD_LONG balances the load between the nodes in the set for you.
```

You can create as many $READ_REPLICA sets as you need. For security and resource isolation, each $READ_REPLICA set has unique connection details.

You use $READ_REPLICA sets for horizontal **read** scaling. To limit data loss for your $SERVICE_LONGs, use [$HA_REPLICAs][ha].

## Prerequisites

To follow this procedure:

- Create a target $SERVICE_LONG.
- Create a [read-only user][read-only-role] on the primary data instance. 

>   A user with the read-only permissions cannot access the primary data instance directly. This user is propagated to the $READ_REPLICA set when you create them.

Reviewer: > A user with the read-only permissions cannot access the primary data instance directly.
This is not correct. All users are created in the primary and propagated to the replicas, including read-only users. A read-only user will not be able to make data changes on to the primary obviously.

Author: Updated

>   A user with the read-only permissions cannot access the primary data instance directly. This user is propagated to the $READ_REPLICA set when you create them.

Reviewer: ```suggestion
  A user with read-only permissions cannot make changes in the primary database. This user is propagated to the $READ_REPLICA set when you create them.
```

Author: Updated

# Manage read replication

You use $READ_REPLICAs to power your read-intensive apps and business intelligence tooling. Using $READ_REPLICAs to serve 
reads for your app removes the load from the primary data instance, and enables your $SERVICE_SHORT to improve ingest performance. 
This is particularly useful when read traffic is very spiky and risks impacting ingest performance, or where reads have 
a lower priority to writes. 

This page shows you how to create and manage $READ_REPLICAs.

## What is read replication?

> A $READ_REPLICAs is a read-only copy of the primary data instance in your $CLOUD_LONG. Queries on $READ_REPLICAs have minimal impact on the performance of the primary data instance. This enables you to interact with 

Reviewer 2: ```suggestion
A $READ_REPLICA is a read-only copy of the primary data instance in your $SERVICE_LONG. Queries on $READ_REPLICAs have minimal impact on the performance of the primary data instance. This enables you to interact with 
```

> up-to-date production data for analysis or to scale out reads beyond the limits of your primary data instance. $READ_REPLICA_CAPs can be short-lived and deleted when a session of data analysis is complete, or long-running to power a

Reviewer 2: $READ_REPLICA_CAPs is not working

> up-to-date production data for analysis or to scale out reads beyond the limits of your primary data instance. Read replicas can be short-lived and deleted when a session of data analysis is complete, or long-running to power a
> business intelligence (BI) tool. 

Reviewer 3: both use cases are about data analysis. maybe the second example could be about serving app servers instead?

```suggestion
or long-running to serve read traffic from your application servers.
```

You use $READ_REPLICAs for **read** scaling. To limit data loss for your $SERVICE_LONGs, use [$HA_REPLICAs][ha].

> You create a $READ_REPLICA as a set of one or more nodes that share the same endpoint in $CONSOLE. You query each set as a single replica, and $CLOUD_LONG balances the load between the nodes within the set for you. 

Reviewer 2: ```suggestion
A $READ_REPLICA is the set of one or more nodes that share the same endpoint in $CONSOLE. You query each set as a single replica. $CLOUD_LONG balances the load between the nodes in the set for you. 
```

> You can create as many $READ_REPLICAs as you need. For security and resource isolation, each $READ_REPLICA has unique connection details. To restrict access without isolation, create a [read-only role][read-only-role] for each $SERVICE_LONG. Users with read-only permissions cannot access the primary data instance directly.

Reviewer 2: Not sure about the link, best to have it in the prereqs only.

## Prerequisites

To follow this procedure:

- Create a target $SERVICE_LONG.
- Create a [read-only user][read-only-role] on the primary data instance. 
  

>    This user is propagated to the $READ_REPLICA when you create it.

Reviewer 2: ```suggestion
   This user is propagated to the $READ_REPLICAs when you create them.
```
